modules/core/src/quaternion.cpp

```diff
@@ -0,0 +1,742 @@
+{	
+    switch (n) {
+        case 0:
+            return w;
+            break;
+        case 1:
+            return x;
+            break;
+        case 2:
+            return y;
+            break;
+        case 3:
+            return z;
+            break;
+        default:
+            throw ("subscript exceeds the index range");
+    }
+}
+
+template <typename T>
+std::ostream & cv::operator<<(std::ostream &os, const Quat<T> &q)
+{
+    os << "Quat " << cv::Vec<T, 4>{q.w, q.x, q.y, q.z};
+    return os;
+}
+
+template <typename T>
+inline Quat<T> Quat<T>::conjugate() const
+{
+    return Quat<T>(w, -x, -y, -z);
+}
+
+template <typename T>
+inline T Quat<T>::norm() const
+{
+    return std::sqrt(dot(*this));
+}
+
+/*
+template <typename T>
+Quat<T> cv::exp(const Quat<T> &q)
+{
+    return q.exp();
+}
+*/
+
+template <typename T>
+Quat<T> Quat<T>::exp() const
+{
+    cv::Vec<T, 3> v{x, y, z};
+    T normV = std::sqrt(v.dot(v));
+    T k = normV < EPS ? 1 : std::sin(normV) / normV;
+    return std::exp(w) * Quat<T>(std::cos(normV), v[0] * k, v[1] * k, v[2] * k);
+}
+
+/*
+template <typename T>
+Quat<T> cv::log(const Quat<T> &q)
+{
+    return q.log();
+}
+*/		
+
+template <typename T>
+Quat<T> Quat<T>::log() const
+{
+    cv::Vec<T, 3> v{x, y, z};
+    T vNorm = std::sqrt(v.dot(v));
+    T qNorm = norm();
+    T k = vNorm < EPS ? 1 : std::acos(w / qNorm) / vNorm;
+    return Quat<T>(std::log(qNorm), v[0] * k, v[1] * k, v[2] *k);
+}
+
+/*
+template <typename T>
+inline Quat<T> cv::power(const Quat<T> &q1, T x)
+{
+    return q1.power(x);
+}
+*/
+
+template <typename T>
+inline Quat<T> Quat<T>::power(T x1) const
+{
+    T angle = getAngle();
+    cv::Vec<T, 3> axis = getAxis();
+    return std::pow(norm(), x1) * Quat<T>(x1 * angle, axis);
+}
+
+
+template <typename T>
+inline Quat<T> sqrt(Quat<T> &q)
+{
+    return q.sqrt();
+}
+
+template <typename T>
+inline Quat<T> Quat<T>::sqrt() const
+{
+    return power(0.5);
+}
+
+/*
+template <typename T>
+inline Quat<T> cv::power(const Quat<T> &p, const Quat<T> &q)
+{
+    return p.power(q);
+}
+*/
+
+template <typename T>
+inline Quat<T> Quat<T>::power(const Quat<T> &q) const
+{
+    Quat<T> ans = *this * q.log();
+    return ans.exp();
+}
+
+template <typename T>
+inline T Quat<T>::dot(Quat<T> q1) const
+{
+    return w * q1.w + x * q1.x + y * q1.y + z * q1.z;
+}
+
+/*
+template <typename T>
+inline Quat<T> cv::crossProduct(const Quat<T> &p, const Quat<T> &q)
+{
+    return p.crossProduct(q);
+}
+*/
+
+template <typename T>
+inline Quat<T> Quat<T>::crossProduct(const Quat<T> &q) const
+{
+    return (*this * q - q * *this) / 2;
+}
+
+template <typename T>
+inline Quat<T> Quat<T>::normalize() const
+{
+    return Quat<T>(w / norm(), x / norm(), y / norm(), z / norm()) ;
+}
+/*
+template <typename T>
+inline Quat<T> cv::inv(const Quat<T> &q)
+{
+    return q.inv();
+}
+*/
+
+template <typename T>
+inline Quat<T> Quat<T>::inv() const
+{
+    return conjugate() / pow(norm(), 2);
```

Comment:
`dot(*this)` is better than `pow(norm(), 2)`: square root then raised to the square.